### ISSUE TITLE
Error message:  "undefined method `empty?'"

### ISSUE THREAD

**Issue by philtech-oh** [TraGicCode/tragiccode-azure_key_vault#42]
Hi tragicode, I am using your module with Puppet 5.5.8 to extract an SSL keypair from Azure Key vault, and getting this error:
```Error: Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: Evaluation Error: Error while evaluating a Function Call, undefined method `empty?' for #<URI::HTTP:0x1852662a> (file: /etc/puppetlabs/code/environments/keyvault_test/manifests/stnint-web.pp, line: 42, column: 16) on node stnint-web00.xxxxx```
The code identified is this:
```  file { '/etc/pki/tls/webkeys/int.www.stxxxxxxx.com.crt':
    content => azure_key_vault::secret('ASC-HFC-KEYVAULT', 'int-www-stxxxxxxx-com-crt', {
      metadata_api_version  => '2018-04-02',
      vault_api_version     => '2016-10-01',
    }),
    ensure  => file,
  }
```
and this is my `hiera.yaml` file:
```
---
version: 5

defaults:
  datadir: data
  data_hash: yaml_data

hierarchy:
  - name: 'Per-node data'
    path: "nodes/%{trusted.certname}.yaml"

  - name: 'Per business unit (silo + tier) data'
    path: "bizunits/%{facts.business_unit}.yaml"

  - name: 'Azure Key Vault Secrets'
    lookup_key: azure_key_vault::lookup
    options:
      vault_name: ASC-HFC-KEYVAULT
      vault_api_version: '2016-10-01'
      metadata_api_version: '2018-04-02'

  - name: 'Common data'
    glob: '*.yaml'```
What is causing this?  This is an opaque error message, but it seems to be from the `secrets` function.

Thanks in advance for your assistance.

**Comment by TraGicCode** [comment]
Hey @philtech-oh,

Thanks for opening an issue.  I noticed you pasted 2 snippets.  1 with your calling the puppet 4 function and another with you showing your setup of hiera.  Recently users that have been using this module have been getting confused between the difference in hiera and puppet 4 function and have been setting up both in their code base ( like you ) without fully understanding if they want to use hiera. So first question.

1. Are you using Hiera lookups in your code anywhere or expect to use secret lookup using hiera?
2. Can you please remove the azure key vault configuration from your hiera.yaml and see if the error goes away?

**NOTE: To be honest i would remove the hiera configuration from your hiera.yaml.  It got added recently and still has some major issues that needs to be fixed.**

**Comment by philtech-oh** [comment]
Removing the azure keyvault from `hiera.yaml` I get the exact same error message (other than the hex value) on the exact same line of code.

More broadly, I'm confused myself between the hiera config and the function.  I don't have specific plans to rely on hiera for secret lookup.

**Comment by TraGicCode** [comment]
> More broadly, I'm confused myself between the hiera config and the function. I don't have specific plans to rely on hiera for secret lookup.

Okay then leave the azure key vault out of your hiera.yaml configure then.

The certification your trying to pull is it stored as a "secret" in your azure key vault?

**Comment by philtech-oh** [comment]
Yes, I have a certificate and a secret key stored in the key vault (this snippet shows the cert):

`az keyvault secret show --vault-name "ASC-HFC-KEYVAULT" -n 'int-www-stxxxxxxx-com-crt' -o table`

```ContentType                  Value
---------------------------  ----------------------------------------------------------------
STN integration certificate  -----BEGIN CERTIFICATE-----
                                        [omitted]
                             -----END CERTIFICATE-----

**Comment by philtech-oh** [comment]
Client:

**Comment by TraGicCode** [comment]
Hey 

Im still not sure what’s going on here. If u look at your error message it looks like it’s saying there is no method called empty? On an URI object.  The strange thing is if you look in the ruby code for this module .empty? Is called but called for secretversion which should only ever be a string.  

Can you show me the puppetserver logs in debug level logging when compiling this catalog?

**Comment by TraGicCode** [comment]
Are you sure that exact code snippet is causing catalog compilation to fail? I can’t fathom how this is happening based on your example

**Comment by TraGicCode** [comment]
The issue is that puppetserver 5.x uses jruby and ruby v1.9.  In this version `Net::HTTP::Get.new(uri)` expects a string and not an object of type URI.  URI's in this version of ruby do not have a .empty? function.  To make this backwards compatible i'm going to convert the uri to a string
